### ISSUE TITLE
Adding dependencies takes forever

### ISSUE THREAD

**Issue by loopmode** [chentsulin/electron-react-boilerplate#1191]
Right now, when you add just about anything via yarn or npm, the process takes incredibly long to finish.

Using node@8.2.1 and yarn@0.27.5, `yarn add video-react` took 406s..:

```
yarn add video-react
yarn add v0.27.5
[1/4] Resolving packages...
[2/4] Fetching packages...
warning 7zip-bin-linux@1.1.0: The platform "win32" is incompatible with this module.
info "7zip-bin-linux@1.1.0" is an optional dependency and failed compatibility check. Excluding it from installation.
warning 7zip-bin-mac@1.0.1: The platform "win32" is incompatible with this module.
info "7zip-bin-mac@1.0.1" is an optional dependency and failed compatibility check. Excluding it from installation.
warning fsevents@1.1.1: The platform "win32" is incompatible with this module.
info "fsevents@1.1.1" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
warning "extract-text-webpack-plugin@2.1.0" has incorrect peer dependency "webpack@^2.2.0".
warning "webpack-dev-server@2.5.0" has incorrect peer dependency "webpack@^2.2.0".
warning "webpack-dev-middleware@1.10.2" has incorrect peer dependency "webpack@1 || ^2.1.0-beta || ^2.2.0-rc.0".
[4/4] Building fresh packages...
success Saved lockfile.
success Saved 3 new dependencies.
├─ classnames@2.2.5
├─ lodash.throttle@4.1.1
└─ video-react@0.8.6
$ concurrently "npm run flow-typed" "npm run build-dll" "electron-builder install-app-deps" "node node_modules/fbjs-scripts/node/check-dev-engines.js package.json"
[2] electron-builder 19.8.0
[2] Installing app dependencies for arch x64 to D:\Projects\loopmode@github\electron-react-video\app
[0]
[0] > electron-react-boilerplate@0.12.0 flow-typed D:\Projects\loopmode@github\electron-react-video
[0] > rimraf flow-typed/npm && flow-typed install --overwrite || true
[0]
[1]
[1] > electron-react-boilerplate@0.12.0 build-dll D:\Projects\loopmode@github\electron-react-video
[1] > cross-env NODE_ENV=development node --trace-warnings -r babel-register ./node_modules/webpack/bin/webpack --config webpack.config.renderer.dev.dll.js --colors
[1]
[3] node node_modules/fbjs-scripts/node/check-dev-engines.js package.json exited with code 0
[0] • Found 86 dependencies in package.json to install libdefs for. Searching...
[0] • rebasing flow-typed cache...
[2] electron-builder install-app-deps exited with code 0
[1] Hash: 8aea1d040f04b551b1dc
[1] Version: webpack 3.0.0
[1] Time: 5577ms
[1]               Asset    Size  Chunks                    Chunk Names
[1] renderer.dev.dll.js  1.7 MB       0  [emitted]  [big]  renderer
[1] [./node_modules/webpack/buildin/module.js] (webpack)/buildin/module.js 517 bytes {0} [built]
[1]    [2] dll renderer 12 bytes {0} [built]
[1]     + 372 hidden modules
[1] npm run build-dll exited with code 0
[0] • Installing 10 libDefs...
[0]   • flow-bin_v0.x.x.js
[0]     └> .\flow-typed\npm\flow-bin_v0.x.x.js
[0]   • redux_v3.x.x.js
[0]     └> .\flow-typed\npm\redux_v3.x.x.js
[0]   • enzyme_v2.3.x.js
[0]     └> .\flow-typed\npm\enzyme_v2.3.x.js
[0]   • express_v4.x.x.js
[0]     └> .\flow-typed\npm\express_v4.x.x.js
[0]   • react-redux_v5.x.x.js
[0]     └> .\flow-typed\npm\react-redux_v5.x.x.js
[0]   • minimist_v1.x.x.js
[0]     └> .\flow-typed\npm\minimist_v1.x.x.js
[0]   • react-router-dom_v4.x.x.js
[0]     └> .\flow-typed\npm\react-router-dom_v4.x.x.js
[0]   • react-router_v4.x.x.js
[0]     └> .\flow-typed\npm\react-router_v4.x.x.js
[0]   • jest_v20.x.x.js
[0]     └> .\flow-typed\npm\jest_v20.x.x.js
[0]   • react-addons-test-utils_v15.x.x.js
[0]     └> .\flow-typed\npm\react-addons-test-utils_v15.x.x.js
[0] • Generating stubs for untyped dependencies...
[0]   • eslint-formatter-pretty@^1.1.0
[0]     └> flow-typed\npm\eslint-formatter-pretty_vx.x.x.js
[0]   • source-map-support@^0.4.15
[0]     └> flow-typed\npm\source-map-support_vx.x.x.js
[0]   • file-loader@^0.11.1
[0]     └> flow-typed\npm\file-loader_vx.x.x.js
[0]   • babel-plugin-dev-expression@^0.2.1
[0]     └> flow-typed\npm\babel-plugin-dev-expression_vx.x.x.js
[0]   • electron-debug@^1.2.0
[0]     └> flow-typed\npm\electron-debug_vx.x.x.js
[0]   • chalk@^2.0.1
[0]     └> flow-typed\npm\chalk_vx.x.x.js
[0]   • stylelint-config-standard@^16.0.0
[0]     └> flow-typed\npm\stylelint-config-standard_vx.x.x.js
[0]   • eslint-import-resolver-webpack@^0.8.3
[0]     └> flow-typed\npm\eslint-import-resolver-webpack_vx.x.x.js
[0]   • eslint-plugin-prettier@^2.2.0
[0]     └> flow-typed\npm\eslint-plugin-prettier_vx.x.x.js
[0]   • rimraf@^2.6.1
[0]     └> flow-typed\npm\rimraf_vx.x.x.js
[0]   • url-loader@^0.5.8
[0]     └> flow-typed\npm\url-loader_vx.x.x.js
[0]   • font-awesome@^4.7.0
[0]     └> flow-typed\npm\font-awesome_vx.x.x.js
[0]   • babel-eslint@^7.2.3
[0]     └> flow-typed\npm\babel-eslint_vx.x.x.js
[0]   • babel-jest@^20.0.3
[0]     └> flow-typed\npm\babel-jest_vx.x.x.js
[0]   • redux-thunk@^2.2.0
[0]     └> flow-typed\npm\redux-thunk_vx.x.x.js
[0]   • history@^4.6.3
[0]     └> flow-typed\npm\history_vx.x.x.js
[0]   • babel-plugin-transform-class-properties@^6.24.1
[0]     └> flow-typed\npm\babel-plugin-transform-class-properties_vx.x.x.js
[0]   • react-router-redux@^5.0.0-alpha.6
[0]     └> flow-typed\npm\react-router-redux_vx.x.x.js
[0]   • babel-preset-env@^1.5.1
[0]     └> flow-typed\npm\babel-preset-env_vx.x.x.js
[0]   • babel-preset-react-hmre@^1.1.1
[0]     └> flow-typed\npm\babel-preset-react-hmre_vx.x.x.js
[0]   • babel-preset-stage-0@^6.24.1
[0]     └> flow-typed\npm\babel-preset-stage-0_vx.x.x.js
[0]   • babel-preset-react-optimize@^1.0.1
[0]     └> flow-typed\npm\babel-preset-react-optimize_vx.x.x.js
[0]   • babel-plugin-dynamic-import-webpack@^1.0.1
[0]     └> flow-typed\npm\babel-plugin-dynamic-import-webpack_vx.x.x.js
[0]   • babel-plugin-add-module-exports@^0.2.1
[0]     └> flow-typed\npm\babel-plugin-add-module-exports_vx.x.x.js
[0]   • enzyme-to-json@^1.5.1
[0]     └> flow-typed\npm\enzyme-to-json_vx.x.x.js
[0]   • electron-devtools-installer@^2.2.0
[0]     └> flow-typed\npm\electron-devtools-installer_vx.x.x.js
[0]   • babel-register@^6.24.1
[0]     └> flow-typed\npm\babel-register_vx.x.x.js
[0]   • babel-preset-react@^6.24.1
[0]     └> flow-typed\npm\babel-preset-react_vx.x.x.js
[0]   • eslint-config-prettier@^2.4.0
[0]     └> flow-typed\npm\eslint-config-prettier_vx.x.x.js
[0]   • css-loader@^0.28.3
[0]     └> flow-typed\npm\css-loader_vx.x.x.js
[0]   • eslint-config-airbnb@^15.0.1
[0]     └> flow-typed\npm\eslint-config-airbnb_vx.x.x.js
[0]   • html-webpack-plugin@^2.29.0
[0]     └> flow-typed\npm\html-webpack-plugin_vx.x.x.js
[0]   • extract-text-webpack-plugin@2.1.0
[0]     └> flow-typed\npm\extract-text-webpack-plugin_vx.x.x.js
[0]   • babili-webpack-plugin@^0.1.2
[0]     └> flow-typed\npm\babili-webpack-plugin_vx.x.x.js
[0]   • redux-logger@^3.0.6
[0]     └> flow-typed\npm\redux-logger_vx.x.x.js
[0]   • prettier@^1.6.1
[0]     └> flow-typed\npm\prettier_vx.x.x.js
[0]   • eslint-plugin-flowtype-errors@^3.3.0
[0]     └> flow-typed\npm\eslint-plugin-flowtype-errors_vx.x.x.js
[0]   • style-loader@^0.18.1
[0]     └> flow-typed\npm\style-loader_vx.x.x.js
[0]   • spectron@^3.7.0
[0]     └> flow-typed\npm\spectron_vx.x.x.js
[0]   • sass-loader@^6.0.6
[0]     └> flow-typed\npm\sass-loader_vx.x.x.js
[0]   • webpack-merge@^4.1.0
[0]     └> flow-typed\npm\webpack-merge_vx.x.x.js
[0]   • stylefmt@^6.0.0
[0]     └> flow-typed\npm\stylefmt_vx.x.x.js
[0]   • cross-spawn@^5.1.0
[0]     └> flow-typed\npm\cross-spawn_vx.x.x.js
[0]   • react-hot-loader@3.0.0-beta.6
[0]     └> flow-typed\npm\react-hot-loader_vx.x.x.js
[0]   • babel-plugin-transform-es2015-classes@^6.24.1
[0]     └> flow-typed\npm\babel-plugin-transform-es2015-classes_vx.x.x.js
[0]   • concurrently@^3.5.0
[0]     └> flow-typed\npm\concurrently_vx.x.x.js
[0]   • electron@^1.6.10
[0]     └> flow-typed\npm\electron_vx.x.x.js
[0]   • eslint-plugin-compat@^1.0.4
[0]     └> flow-typed\npm\eslint-plugin-compat_vx.x.x.js
[0]   • eslint-plugin-jest@^20.0.3
[0]     └> flow-typed\npm\eslint-plugin-jest_vx.x.x.js
[0]   • cross-env@^5.0.0
[0]     └> flow-typed\npm\cross-env_vx.x.x.js
[0]   • babel-loader@^7.1.0
[0]     └> flow-typed\npm\babel-loader_vx.x.x.js
[0]   • fbjs-scripts@^0.8.0
[0]     └> flow-typed\npm\fbjs-scripts_vx.x.x.js
[0]   • eslint-plugin-promise@^3.5.0
[0]     └> flow-typed\npm\eslint-plugin-promise_vx.x.x.js
[0]   • eslint-plugin-react@^7.1.0
[0]     └> flow-typed\npm\eslint-plugin-react_vx.x.x.js
[0]   • eslint-plugin-import@^2.6.0
[0]     └> flow-typed\npm\eslint-plugin-import_vx.x.x.js
[0]   • identity-obj-proxy@^3.0.0
[0]     └> flow-typed\npm\identity-obj-proxy_vx.x.x.js
[0]   • webpack-bundle-analyzer@^2.8.2
[0]     └> flow-typed\npm\webpack-bundle-analyzer_vx.x.x.js
[0]   • devtron@^1.4.0
[0]     └> flow-typed\npm\devtron_vx.x.x.js
[0]   • react-test-renderer@^15.6.1
[0]     └> flow-typed\npm\react-test-renderer_vx.x.x.js
[0]   • webpack@^3.0.0
[0]     └> flow-typed\npm\webpack_vx.x.x.js
[0]   • flow-typed@^2.1.2
[0]     └> flow-typed\npm\flow-typed_vx.x.x.js
[0]   • eslint-plugin-jsx-a11y@5.0.3
[0]     └> flow-typed\npm\eslint-plugin-jsx-a11y_vx.x.x.js
[0]   • eslint@^3.19.0
[0]     └> flow-typed\npm\eslint_vx.x.x.js
[0]   • stylelint@^7.12.0
[0]     └> flow-typed\npm\stylelint_vx.x.x.js
[0]   • webpack-dev-server@^2.5.0
[0]     └> flow-typed\npm\webpack-dev-server_vx.x.x.js
[0]   • eslint-plugin-flowtype@^2.33.0
[0]     └> flow-typed\npm\eslint-plugin-flowtype_vx.x.x.js
[0]   • babel-core@^6.24.1
[0]     └> flow-typed\npm\babel-core_vx.x.x.js
[0]   • video-react@^0.8.6
[0]     └> flow-typed\npm\video-react_vx.x.x.js
[0]   • flow-runtime@^0.13.0
[0]     └> flow-typed\npm\flow-runtime_vx.x.x.js
[0]   • electron-builder@^19.8.0
[0]     └> flow-typed\npm\electron-builder_vx.x.x.js
[0]   • sinon@^2.3.5
[0]     └> flow-typed\npm\sinon_vx.x.x.js
[0]   • node-sass@^4.5.3
[0]     └> flow-typed\npm\node-sass_vx.x.x.js
[0]   • jsdom@^11.0.0
[0]     └> flow-typed\npm\jsdom_vx.x.x.js
[0]   • babel-plugin-flow-runtime@^0.11.1
[0]     └> flow-typed\npm\babel-plugin-flow-runtime_vx.x.x.js
[0]
[0] !! No flow@v0.48.0-compatible libdefs found in flow-typed for the above untyped dependencies !!
[0]
[0] I've generated `any`-typed stubs for these packages, but consider submitting
[0] libdefs for them to https://github.com/flowtype/flow-typed/
[0]
[0] npm run flow-typed exited with code 0
Done in 406.58s.
```

In this case for example, `classnames` was already added by `video-react`.
However, just doing `yarn add classnames` in order to have it as an explicit dependency in `package.json` did the whole thing again, taking another 415s. That's almost 7 minutes! [Ain't nobody got time for that!](https://www.youtube.com/watch?v=6gLMSf4afzo)

What takes up most of the time seems to be `Installing 10 libDefs`.
At least, `[1] npm run build-dll exited with code 0` is reached pretty quickly and the CLI seems to stall from there until the libDefs report is printed.

I guess it's about native libs. And it should be possible to omit the step, at least it shouldn't be always performed implicitly, but rather explicitly and only on-demand.
I understand that it's absolutely necessary the first time a user installs the project.

A couple of ideas:

- Offer a custom install/setup script that supports something like a `--native` flag
- Or only perform it when there is not e.g. a `.nativelibs` file in the project, that gets generated. One could force the user to delete the file so that the next install would perform the full setup again and create a new `.nativelibs` file
- And probably there are many other ways, maybe even a standard one? (I'm totally new to electron)

**Comment by Slapbox** [comment]
You can add `--ignoreScripts` if you have a few dependencies to add, but on the last run you'll of coruse need to not ignore scripts.

**Comment by loopmode** [comment]
Well that is good to know, thanks. Especially useful when you install multiple dependencies - which, however, is rather rare.
If you find the time (or if there already is a resource somewhere), I'd be interested in what that step does and why it takes so long. What's going on there?

In another project, I also had terrible durations on install, and it turned out to be `material-design-icons` that takes forever - and since we only needed a bunch of icons, we manually copied them to the project and lost the dependency, cutting over a minute from the setup duration.
Maybe something similar can be done here, maybe it's something different, but it surely can be optimized?

**Comment by Slapbox** [comment]
Do you have any native dependencies? For example, I use `electron-spell-checker` (I think is the name) and it needs to rebuild with node-gyp on every install.

Also worth asking about your HDD. I was running on a slow SSD and it was taking a long time on installs with scripts. Moved to a faster SSD and the longest I wait is about 130 seconds now. Still a time sink, but much less so than 600 seconds.

**Comment by amilajack** [comment]
Should be fixed when `0.13.0` lands
